lib/group-update.js

```diff
@@ -0,0 +1,335 @@
+                            resultlist.rows.forEach(element => {
+                                students.push(element.user_list);
+                            });
+                            module.exports._shuffle(students);
+                            var num_students = resultlist.rowCount;
+                            var not_assigned = [];
+                            const resultlist2 = await sqldb.queryAsync(sql.select_not_assigned, aid);
+                            resultlist2.rows.forEach(element => {
+                                not_assigned.push(element.user_list);
+                            });
+                            module.exports._shuffle(not_assigned);
+                            var num_not_assigned = resultlist2.rowCount;
+                            job.verbose(`There are ` + num_students + ' students enrolled in ' + assessment_label);
+                            job.verbose(num_not_assigned + ' of them have not been in a group');
+                            job.verbose(`----------------------------------------`);
+                            job.verbose(`Processing creating groups - max of ` + maxnum + ' and min of ' + minnum);
+                            let num_group = Math.ceil(num_students / maxnum);
+                            let successCount = 0;
+                            let errorCount = 0;
+                            let output = null;
+                            let outputCount = 0;
+                            let outputThreshold = 100;
+                            for(let i = 0; i < num_group; i++){
+                                //fill in the group
+                                let teamname = 'team' + i;
+                                let members = [];
+                                for(let j = 0; j < maxnum; j++){
+                                    if(students.length > 0){
+                                        let uid = students.pop();
+                                        members.push(uid);
+                                        job.verbose(uid);                                        
+                                    }
+                                }
+                                for(const member of members){
+                                    const msg = 'Processing record ' + member;
+                                    if (output == null) {
+                                        output = msg;
+                                    } else {
+                                        output += '\n' + msg;
+                                    }
+                                    try{
+                                        const params = [
+                                            assessment_id,
+                                            teamname,
+                                            member,
+                                        ];
+                                        await sqldb.callAsync('assessment_groups_update', params);
```

Comment:
It is going to be very slow to make all these separate DB calls. It would be better to batch up all the new group information and make one DB call that creates all the new groups. As an example, see https://github.com/PrairieLearn/PrairieLearn/blob/master/sprocs/sync_news_items.sql